packages/react-composites/src/composites/ChatComposite/adapter/AzureCommunicationChatAdapter.test.ts

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { ChatClientState, ChatThreadClientState } from 'chat-stateful-client';
+import { waitWithBreakCondition } from '../../../mocks';
+import {
+  ChatThreadClientOverrides,
+  createMockStatefulChatClient,
+  StatefulChatClientOverrides
+} from '../../../mocks/ChatMocks';
+import { AzureCommunicationChatAdapter } from './AzureCommunicationChatAdapter';
+
+const MOCK_THREADID = 'mockThreadId';
+const SEND_ERROR = 'send error';
+const UNSUBSCRIBE_ERROR = 'unsubscribe error';
+const RECEIPT_ERROR = 'receipt error';
+const TYPING_ERROR = 'typing error';
+const REMOVE_ERROR = 'remove error';
+const TOPIC_ERROR = 'topic error';
+
+const THREAD_OVERRIDES: ChatThreadClientOverrides = {
+  threadId: MOCK_THREADID,
+  sendMessage: () => {
+    throw new Error(SEND_ERROR);
+  },
+  sendReadReceipt: () => {
+    throw new Error(RECEIPT_ERROR);
+  },
+  sendTypingNotification: () => {
+    throw new Error(TYPING_ERROR);
+  },
+  removeParticipant: () => {
+    throw new Error(REMOVE_ERROR);
+  },
+  updateTopic: () => {
+    throw new Error(TOPIC_ERROR);
+  }
+};
+
+const STATEFUL_OVERRIDES: StatefulChatClientOverrides = {
+  getState: () => {
+    const threads = new Map<string, ChatThreadClientState>();
+    threads.set(MOCK_THREADID, {} as ChatThreadClientState);
+    return {
+      userId: { kind: 'communicationUser', communicationUserId: '' },
+      threads
+    } as ChatClientState;
+  },
+  off: () => {
+    throw new Error(UNSUBSCRIBE_ERROR);
+  }
+};
+
+describe('AzureCommunicationChatAdapter', () => {
+  test('emits error event when error happens', async () => {
+    let error: Error | undefined;
+    const errorListener = (e: Error): void => {
+      error = e;
+    };
+
+    const statefulChatClient = createMockStatefulChatClient(STATEFUL_OVERRIDES, THREAD_OVERRIDES);
+    const chatThreadClient = statefulChatClient.getChatThreadClient('');
+    const adapter: AzureCommunicationChatAdapter = new AzureCommunicationChatAdapter(
+      statefulChatClient,
+      chatThreadClient
+    );
+
+    adapter.on('error', errorListener);
+
+    adapter.dispose();
+
+    await waitWithBreakCondition(() => error !== undefined);
```

Comment:
We should really avoid doing this as it adds large amounts of arbitrary time to unit tests. Whats the waiting for? Is there a way we can use jest mock timers to speed up triggering the wait?

Reply:
It is because the dispose() runs some asynchronous code but dispose doesn't return a promise to await on. The waitWithBreakCondition is a way to wait so the asynchronous code in dispose() completes before we validate it in the test. In practice this is almost immediate since it is mocked out but still asynchronous. The waitWithBreakCondition waits for 100ms at a time in a loop. On my PC it only waits once before we can continue on with the test verification. We can shorten this 100ms even more if we want to do that. Another thing we can do is make adapter's dispose() return a promise but this changes the public API so will be a breaking change.

Reply:
If the adapter dispose is itself firing off awaitable async functions that we are not awaiting we should definitely update this to be a `Promise<void>` regardless of the breaking change

Reply:
Taking a look presumably this will simply become:
```
  public dispose(): Promise<void> {
    this.callClient.offStateChange(this.onClientStateChange);
    await this.callAgent.dispose();
  }
```

Reply:
Right, that looks to be the change we'll need to make it awaitable.

Reply:
Updated this so that CallAdapter's dispose function returns a Promise. The ChatAdapter's dispose function is unchanged (still return void) since there is nothing to await. Also removed the waitWithBreakCondition from both tests.